Extensions/MUC-Sub/XMPPMUCSub.m

```diff
@@ -0,0 +1,699 @@
+        }
+    }};
+    
+    if (dispatch_get_specific(moduleQueueTag)) {
+        block();
+    }
+    else {
+        dispatch_async(moduleQueue, block);
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+#pragma mark -
+#pragma mark Message + Presence Delegate
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+- (BOOL)xmppStream:(XMPPStream *)sender didReceiveIQ:(XMPPIQ *)iq
+{
+    NSString *type = iq.type;
+    
+    if ([type isEqualToString:@"result"] || [type isEqualToString:@"error"]) {
+        return [xmppIDTracker invokeForID:iq.elementID withObject:iq];
+    }
+    
+    return NO;
+}
+
+
+- (void)xmppStream:(XMPPStream *)sender didReceiveMessage:(XMPPMessage *)message
+{
+    // <message from="coven@muc.shakespeare.example"
+    //            to="hag66@shakespeare.example/pda">
+    //   <event xmlns="http://jabber.org/protocol/pubsub#event">
+    //     <items node="urn:xmpp:mucsub:nodes:messages">
+    //       <item id="18277869892147515942">
+    //         <message from="coven@muc.shakespeare.example/secondwitch"
+    //                    to="hag66@shakespeare.example/pda"
+    //                  type="groupchat"
+    //                 xmlns="jabber:client">
+    //           <archived xmlns="urn:xmpp:mam:tmp"
+    //                        by="muc.shakespeare.example"
+    //                        id="1467896732929849" />
+    //           <stanza-id xmlns="urn:xmpp:sid:0"
+    //                         by="muc.shakespeare.example"
+    //                         id="1467896732929849" />
+    //           <body>Hello from the MUC room !</body>
+    //         </message>
+    //       </item>
+    //     </items>
+    //   </event>
+    // </message>
+    
+    NSXMLElement* items = [self findMUCSubItemsElement:message forEvent:@"messages"];
+    if (nil == items) {
+        return;
+    }
+    
+    // All preconditions show that it's a MUC-Sub message. Extract the original message
+    // and forward it to the registered delegates. The message may contain several <item>
+    // elements and thus several original <message> elements.
+    dispatch_block_t block = ^{ @autoreleasepool {
+        for (NSXMLNode *item in items.children) {
+            NSXMLNode* messageNode = [item childAtIndex:0];
+            if (nil == messageNode) {
+                continue;
+            }
+            
+            if (NSXMLElementKind != messageNode.kind) {
+                continue;
+            }
+            
+            XMPPMessage *m = [XMPPMessage messageFromElement:(NSXMLElement *)messageNode];
+            [multicastDelegate xmppStream:sender didReceiveMessage:m];
+        }
+    }};
+    
+    if (dispatch_get_specific(moduleQueueTag)) {
+        block();
+    }
+    else {
+        dispatch_async(moduleQueue, block);
```

Comment:
moduleQueue is serial, right?

Reply:
It's either the queue that is provided on initialization or the initializer of `XMPPModule` creates one.
 ```swift
const char *moduleQueueName = [[self moduleName] UTF8String];
moduleQueue = dispatch_queue_create(moduleQueueName, NULL);
```